File: backend/src/main/java/dat250/FeedApp/manager/FeedAppManager.java
package dat250.FeedApp.manager;

import dat250.FeedApp.controller.PollController;
import dat250.FeedApp.controller.VoteController;
import dat250.FeedApp.domain.Poll;
import dat250.FeedApp.domain.User;
import dat250.FeedApp.domain.Vote;
import dat250.FeedApp.domain.VoteOption;
import dat250.FeedApp.domain.VoteOptionCount;
import dat250.FeedApp.repository.PollRepository;
import dat250.FeedApp.repository.UserRepository;
import dat250.FeedApp.repository.VoteOptionRepository;
import dat250.FeedApp.repository.VoteRepository;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class FeedAppManager {
	private UserRepository userRepo;
	private PollRepository pollRepo;
	private VoteOptionRepository voteOptionRepo;
	private VoteRepository voteRepo;

	FeedAppManager(@Autowired UserRepository userRepository, @Autowired PollRepository pollRepository, @Autowired VoteOptionRepository voteOptionRepository, @Autowired VoteRepository voteRepository) {
	this.userRepo = userRepository;
	this.pollRepo = pollRepository;
	this.voteOptionRepo = voteOptionRepository;
	this.voteRepo = voteRepository;
	}

	public Optional<User> findUser(String username) {
		return userRepo.findById(username);
	}

	public User saveUser(User user) {
		return userRepo.save(user);
	}

	public List<Poll> findAllPolls() {
		List<Poll> pollsList = new ArrayList<>();
		pollRepo.findAll().forEach(pollsList::add);
		return pollsList;
	}

	public Poll addPoll(String username, PollController.PollRequest pollRequest) {
		Optional<User> user = findUser(username);
		if (user.isEmpty()) {
			return null;
		}

		Poll poll = pollRepo.save(new Poll(
				pollRequest.question(),
				Instant.now(),
				pollRequest.validUntil(),
				user.get()));

		for (String vo : pollRequest.voteOptions()) {
			voteOptionRepo.save(new VoteOption(vo, poll));
		}

		return poll;
	}

	public void deletePoll(Long pollId) {
		pollRepo.deleteById(pollId);
	}

	public Vote castVote(String username, VoteController.VoteRequest voteRequest) {
		Optional<User> user = findUser(username);
		if (user.isEmpty()) {
			return null;
		}

		VoteOption vo = voteOptionRepo.findByPollIdAndCaption(voteRequest.pollId(), voteRequest.caption());
		if (vo == null) {
			return null;
		}

		return voteRepo.save(new Vote(Instant.now(), vo, user.get()));
	}

	public List<Vote> getVotesForPoll(Long id) {
		// TODO: caching
		return voteRepo.findAllVotesByPollId(id);
	}

	public List<VoteOptionCount> getVoteOptionCountForPoll(Long id) {
		// TODO: caching
		return voteOptionRepo.findVoteOptionCountsByPollId(id);
	}
}

File: backend/src/main/kotlin/dat250/FeedApp/manager/FeedAppManager.kt
package dat250.FeedApp.manager

import dat250.FeedApp.controller.PollController
import dat250.FeedApp.controller.VoteController
import dat250.FeedApp.domain.Poll
import dat250.FeedApp.domain.User
import dat250.FeedApp.domain.Vote
import dat250.FeedApp.domain.VoteOption
import dat250.FeedApp.domain.VoteOptionCount
import dat250.FeedApp.messaging.FeedEventPublisher
import dat250.FeedApp.messaging.PollCreatedEvent
import dat250.FeedApp.messaging.VoteCastEvent
import dat250.FeedApp.repository.PollRepository
import dat250.FeedApp.repository.UserRepository
import dat250.FeedApp.repository.VoteOptionRepository
import dat250.FeedApp.repository.VoteRepository
import java.time.Instant
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.Caching
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class FeedAppManager(
        private val userRepo: UserRepository,
        private val pollRepo: PollRepository,
        private val voteOptionRepo: VoteOptionRepository,
        private val voteRepo: VoteRepository,
        private val eventPublisher: FeedEventPublisher
) {
    fun findUser(username: String): User? = userRepo.findByIdOrNull(username)

    fun saveUser(user: User): User = userRepo.save(user)

    fun findAllPolls(): List<Poll> = pollRepo.findAll().toList()

    fun addPoll(username: String, pollRequest: PollController.PollRequest): Poll? {
        val user = findUser(username)
        if (user == null) {
            return null
        }
        val poll =
                pollRepo.save(
                        Poll(
                                question = pollRequest.question,
                                publishedAt = Instant.now(),
                                validUntil = pollRequest.validUntil,
                                creator = user,
                        )
                )
        for (vo in pollRequest.voteOptions) {
            voteOptionRepo.save(VoteOption(caption = vo, poll = poll))
        }
        eventPublisher.publishPollCreated(
                PollCreatedEvent(
                        pollId = poll.id,
                        question = poll.question,
                        publishedAt = poll.publishedAt,
                        validUntil = poll.validUntil,
                        creator = user.username,
                        voteOptions = pollRequest.voteOptions
                )
        )
        return poll
    }

    @Caching(
            evict = [
                CacheEvict(cacheNames = ["pollVotes"], key = "#pollId"),
                CacheEvict(cacheNames = ["pollVoteCounts"], key = "#pollId")
            ]
    )
    fun deletePoll(pollId: Long) = pollRepo.deleteById(pollId)

    @Caching(
            evict = [
                CacheEvict(cacheNames = ["pollVotes"], key = "#voteRequest.pollId"),
                CacheEvict(cacheNames = ["pollVoteCounts"], key = "#voteRequest.pollId")
            ]
    )
    fun castVote(username: String, voteRequest: VoteController.VoteRequest): Vote? {
        val user = findUser(username)
        if (user == null) {
            return null
        }
        val vo = voteOptionRepo.findByPollIdAndCaption(voteRequest.pollId, voteRequest.caption)
        if (vo == null) {
            return null
        }
        val vote = voteRepo.save(Vote(publishedAt = Instant.now(), voteOption = vo, user = user))
        eventPublisher.publishVoteCast(
                VoteCastEvent(
                        pollId = vo.poll.id,
                        voteOption = vo.caption,
                        voter = user.username,
                        publishedAt = vote.publishedAt
                )
        )
        return vote
    }

    @Cacheable(cacheNames = ["pollVotes"], key = "#id")
    fun getVotesForPoll(id: Long): List<Vote> = voteRepo.findAllVotesByPollId(id)

    @Cacheable(cacheNames = ["pollVoteCounts"], key = "#id")
    fun getVoteOptionCountForPoll(id: Long): List<VoteOptionCount> =
            voteOptionRepo.findVoteOptionCountsByPollId(id)
}
